src/functions/env.ts

```diff
@@ -0,0 +1,238 @@
+  `\\s*'(?:\\'|[^'])*'|` + //     single quoted or
+  `\\s*"(?:\\"|[^"])*"|` + //     double quoted or
+  "[^\\#\\r\\n]+" +        //     unquoted
+  ")?" +                   //   end optional value
+  "\\s*" +                 //   trailing whitespaces
+  "(?:\\#[^\\n]*)?" +      //   optional comment
+  "$",                     // end line
+  "gms"                    // flags: global, multiline, dotall
+);
+
+interface ParseResult {
+  envs: Record<string, string>;
+  errors: string[];
+}
+
+/**
+ * Parse contents of a dotenv file.
+ *
+ * Each line should contain key, value pairs, e.g.:
+ *
+ *   SERVICE_URL=https://example.com
+ *
+ * Values can be double quoted, e.g.:
+ *
+ *   SERVICE_URL="https://example.com"
+ *
+ * Double quoted values can include newlines, e.g.:
+ *
+ *   PUBLIC_KEY="-----BEGIN PUBLIC KEY-----\nABC\nEFG\n-----BEGIN PUBLIC KEY-----""
+ *
+ * or span multiple lines, e.g.:
+ *
+ *   PUBLIC_KEY="-----BEGIN PUBLIC KEY-----
+ *   ABC
+ *   EFG
+ *   -----BEGIN PUBLIC KEY-----"
+ *
+ * See test for more examples.
+ *
+ * @return {ParseResult} Result containing parsed key, value pairs and errored lines.
+ */
+export function parse(data: string): ParseResult {
+  const envs: Record<string, string> = {};
+  const errors: string[] = [];
+
+  data = data.replace(/\r\n?/, "\n"); // For Windows support.
+  let match;
+  while ((match = LINE_RE.exec(data))) {
+    let [, k, v] = match;
+    v = (v || "").trim();
+
+    let quotesMatch;
+    if ((quotesMatch = /^(["'])(.*)\1$/ms.exec(v)) != null) {
+      // Remove surrounding single/double quotes.
+      v = quotesMatch[2];
+      if (quotesMatch[1] === '"') {
+        // Unescape newlines and tabs.
+        v = v.replace("\\n", "\n").replace("\\r", "\r").replace("\\t", "\t").replace("\\v", "\v");
+        // Unescape other escapable characters.
+        v = v.replace(/\\([\\'"])/g, "$1");
+      }
+    }
+    envs[k] = v;
+  }
+
+  const nonmatches = data.replace(LINE_RE, "");
+  for (let line of nonmatches.split(/[\r\n]+/)) {
+    line = line.trim();
+    if (line.startsWith("#")) {
+      // Ignore comments
+      continue;
+    }
+    if (line.length) errors.push(line);
+  }
+
+  return { envs, errors };
+}
+
+/**
+ * Validates string for use as an env var key.
+ *
+ * We restrict key names to ones that conform to POSIX standards.
+ * This is more restrictive than what is allowed in Cloud Functions or Cloud Run.
+ */
+export function validateKey(key: string): void {
+  if (RESERVED_KEYS.includes(key)) {
+    throw new Error("Key reserved for internal use.");
+  }
+  if (!/^[A-Z_][A-Z0-9_]*$/.test(key)) {
+    throw new Error(
+      "Key must start with an uppercase ASCII letter or underscore" +
+        ", and then consist of uppercase ASCII letters, digits, and underscores."
+    );
+  }
+  if (key.startsWith("X_GOOGLE_") || key.startsWith("FIREBASE_")) {
+    throw new Error("Key starts with a reserved prefix (X_GOOGLE_ or FIREBASE_)");
```

Comment:
nit: I would consider using the same error as above so that our error space doesn't strictly map to our implementation.